server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/dataloader/DataReadManager.java

```diff
@@ -106,45 +195,49 @@ public Session generateSession(DataReadRequest request) {
                                 .status(Status.DataStatus.UNPROCESSED)
                                 .build())
                         .collect(Collectors.toList()),
-                1000).forEach(dataReadLogDao::batchInsert);
-
-        return session;
-
+                this.insertBatchSize
+        ).forEach(dataReadLogDao::batchInsert);
+        // save session
+        sessionDao.updateToFinished(session.getId());
     }
 
     /**
      * Assign data for consumer
      *
      * @param consumerId consumer
-     * @param session session
+     * @param sessionId    sessionId
      * @return data
      */
     @Transactional
-    public DataReadLog assignmentData(String consumerId, Session session) {
-        var sid = session.getId();
-        var queue = sessionCache.get(String.valueOf(sid), LinkedList::new);
+    public DataReadLog assignmentData(String consumerId, Long sessionId) {
+        var queue = sessionCache.get(String.valueOf(sessionId), LinkedList::new);
 
         if (queue.isEmpty()) {
-            queue.addAll(dataReadLogDao.selectTopsUnAssignedData(sid, cacheSize));
+            queue.addAll(dataReadLogDao.selectTopsUnAssignedData(sessionId, cacheSize));
         }
         DataReadLog readLog = queue.poll();
         if (Objects.nonNull(readLog)) {
             readLog.setConsumerId(consumerId);
             readLog.setAssignedNum(readLog.getAssignedNum() + 1);
             dataReadLogDao.updateToAssigned(readLog);
             log.info("Assignment data id: {} to consumer:{}", readLog.getId(), readLog.getConsumerId());
+        } else {
+            var session = sessionDao.selectOne(sessionId);
+            if (session.getStatus() == Status.SessionStatus.UNFINISHED) {
+                throw new SwRequestFrequentException(
+                        DATASET_LOAD, "data load: index is building, please try again later");
+            }
```

Comment:
Is it possible for the following abnormal situation to occur?
1. The async generating was not done (more data should be processed).
2. This function is called and gets nothing.
3. The generating is done and mark the session finished.
4. This function check the session, and see the finished session.

Reply:
yes, you are right, already use the origin session object which was passed by outer.